File: WolfDen/wwwroot/ts/Behaviours/CombatBehaviour/NormalCombatBehaviourTemplate.ts
/// <reference path="../../Unit.ts" />

namespace TurnBasedBoardGame {
    export class NormalCombatBehaviourTemplate implements IBehaviourTemplate {
        
        Class: string;
        MaxHealth: number = 0;
        UnitType: UnitType = UnitType.Infantry;
        DamageChart: DamageChart;

        AddBehaviour(unit: Unit): void {
            let self = this;
            let combatBehaviour = new NormalCombatBehaviour();
            Object.assign(combatBehaviour, self);
            let moveBehaviour = null;
            for (let key in unit.Behaviours)
                if (IsIMoveBehaviour(unit.Behaviours[key]))
                    moveBehaviour = unit.Behaviours[key];
            combatBehaviour.MoveBehaviour = moveBehaviour;
            combatBehaviour.Unit = unit;
            combatBehaviour.Game = unit.Game;
            combatBehaviour.CurrentHealth = self.MaxHealth; 
            combatBehaviour.Id = unit.Id + self.Class;
            combatBehaviour.Initialize();
            
            unit.Behaviours[combatBehaviour.Id] = combatBehaviour;
        }
    }
}

File: WolfDen/Game/Behaviours/CombatBehaviour/NormalCombatBehaviourTemplate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using WolfDen.Game;
using WolfDen.Game.Interfaces;
using WolfDen.Game.Templates;

namespace WolfDen.Game.Behaviours.CombatBehaviour
{
    public class NormalCombatBehaviourTemplate : IBehaviourTemplate
    {
        public string Class { get { return GetType().Name; } }
        public int MaxHealth { get; set; }
        public UnitType UnitType { get; set; }
        public DamageChart DamageChart { get; set; } = DamageChart.Instance;

        public virtual void AddBehaviour(Unit unit)
        {
            var moveBehaviour = unit.Behaviours.Values.OfType<IMoveBehaviour>().FirstOrDefault();
            // Note: if the unit has a movebehaviour make sure it was added before this.
            var combatBehaviour = new NormalCombatBehaviour(unit.Game, unit, MaxHealth, UnitType, moveBehaviour, DamageChart);
            combatBehaviour.Id = unit.Id + Class;
            unit.Behaviours.Add(combatBehaviour.Id, combatBehaviour);
        }
    }
}
